fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/AbstractFurnaceBlockEntityMixin.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.item;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.AbstractFurnaceBlockEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.collection.DefaultedList;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+@Mixin(AbstractFurnaceBlockEntity.class)
+public abstract class AbstractFurnaceBlockEntityMixin {
+	@Shadow
+	protected DefaultedList<ItemStack> inventory;
+	private static DefaultedList<ItemStack> capturedInventory;
+
+	@Inject(method = "tick", at = @At("HEAD"))
+	private static void getStackCraftingRemainder(World world, BlockPos pos, BlockState state, AbstractFurnaceBlockEntity blockEntity, CallbackInfo ci) {
+		capturedInventory = ((AbstractFurnaceBlockEntityMixin) (Object) blockEntity).inventory;
+	}
+
+	@Redirect(method = "tick", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;isEmpty()Z", ordinal = 2))
+	private static boolean setStackCraftingRemainder(ItemStack itemStack) {
+		if (itemStack.hasRecipeRemainder()) {
```

Comment:
Needs testing with a furnace smelting wet sponge using multiple buckets. Note that the furnace must be fueled by other fuel first and the buckets must be inserted with command. The code here suggests that the empty bucket is always replaced; the redirected `isEmpty()` check, from what I can see, makes it only replace on the last bucket, but I couldn't reproduce it on vanilla.

Also, `itemStack` may be empty as this is called after `decrement`, which needs documenting. (Decrementing cannot turn the stack into `ItemStack#EMPTY`, however, as `item` field is final.)

Reply:
This is the same behavior as vanilla do.

[](https://user-images.githubusercontent.com/32025549/188224151-c122684e-48d1-4778-895e-f469d504d79f.mp4)

Reply:
Seems to work as intended